JIT: Generalize handling of commas in block morphing

jakobbotsch opened the pull request:
Eliminate commas early in block morphing, before the rest of the transformation
needs to look at it. Do this by extracting their side effects and adding
them on top of the returned result instead. This allows us to handle
arbitrary COMMAs on destination operand in a general manner.

Prerequisite to #83388.

Fix #1699.

jakobbotsch commented:
cc @dotnet/jit-contrib PTAL @AndyAyersMS @SingleAccretion 

There is a small TP cost, however it is not due to the new code I have added. Rather, the detailed TP breakdown looks like:
```scala
Base: 55040568185, Diff: 55058841007, +0.0332%

?fgMorphSmpOpOptional@Compiler@@AEAAPEAUGenTree@@PEAUGenTreeOp@@PEA_N@Z                                                                                                                                                            : 19203600 : +10.94%  : 40.48% : +0.0349%
?Morph@MorphInitBlockHelper@@IEAAPEAUGenTree@@XZ                                                                                                                                                                                   : 9387345  : NA       : 19.79% : +0.0171%
memset                                                                                                                                                                                                                             : 2674938  : +0.60%   : 5.64%  : +0.0049%
?EvalArgsToTemps@CallArgs@@QEAAXPEAVCompiler@@PEAUGenTreeCall@@@Z                                                                                                                                                                  : 798509   : +2.87%   : 1.68%  : +0.0015%
?fgPerBlockLocalVarLiveness@Compiler@@QEAAXXZ                                                                                                                                                                                      : 226273   : +0.08%   : 0.48%  : +0.0004%
GenTreeVisitor<`Compiler::fgMorphStmtBlockOps'::`2'::Visitor>::WalkTree                                                                                                                                                            : 184449   : +15.81%  : 0.39%  : +0.0003%
?fgMakeOutgoingStructArgCopy@Compiler@@AEAAXPEAUGenTreeCall@@PEAVCallArg@@@Z                                                                                                                                                       : 136120   : +18.70%  : 0.29%  : +0.0002%
?newRefPositionRaw@LinearScan@@AEAAPEAVRefPosition@@IPEAUGenTree@@W4RefType@@@Z                                                                                                                                                    : -48040   : -0.01%   : 0.10%  : -0.0001%
?lvaComputeRefCounts@Compiler@@QEAAX_N0@Z                                                                                                                                                                                          : -51153   : -0.01%   : 0.11%  : -0.0001%
?resolveRegisters@LinearScan@@QEAAXXZ                                                                                                                                                                                              : -51409   : -0.01%   : 0.11%  : -0.0001%
?allocateMemory@ArenaAllocator@@QEAAPEAX_K@Z                                                                                                                                                                                       : -61660   : -0.01%   : 0.13%  : -0.0001%
?processBlockStartLocations@LinearScan@@AEAAXPEAUBasicBlock@@@Z                                                                                                                                                                    : -64151   : -0.01%   : 0.14%  : -0.0001%
?optCopyProp@Compiler@@QEAA_NPEAUBasicBlock@@PEAUStatement@@PEAUGenTreeLclVarCommon@@IPEAV?$JitHashTable@IU?$JitSmallPrimitiveKeyFuncs@I@@PEAV?$ArrayStack@VCopyPropSsaDef@Compiler@@@@VCompAllocator@@VJitHashTableBehavior@@@@@Z : -71693   : -0.01%   : 0.15%  : -0.0001%
?InitializeRange@?$JitExpandArray@E@@IEAAXII@Z                                                                                                                                                                                     : -135712  : -0.08%   : 0.29%  : -0.0002%
?allocateRegisters@LinearScan@@QEAAXXZ                                                                                                                                                                                             : -142056  : -0.01%   : 0.30%  : -0.0003%
?allocateReg@LinearScan@@AEAA?AW4_regNumber_enum@@PEAVInterval@@PEAVRefPosition@@@Z                                                                                                                                                : -160379  : -0.02%   : 0.34%  : -0.0003%
?fgMorphInitBlock@Compiler@@QEAAPEAUGenTree@@PEAU2@@Z                                                                                                                                                                              : -569150  : -100.00% : 1.20%  : -0.0010%
?PrepareSrc@MorphCopyBlockHelper@@MEAAXXZ                                                                                                                                                                                          : -886968  : -34.33%  : 1.87%  : -0.0016%
?PrepareDst@MorphInitBlockHelper@@IEAAXXZ                                                                                                                                                                                          : -4393789 : -100.00% : 9.26%  : -0.0080%
?fgMorphCopyBlock@Compiler@@QEAAPEAUGenTree@@PEAU2@@Z                                                                                                                                                                              : -5063084 : -100.00% : 10.67% : -0.0092%
```

What actually happens is that MSVC changes inlining strategy for `fgMorphCopyBlock` and `fgMorphInitBlock` called by `fgMorphSmpOpOptional`, which leads to some missed GDVs. Some pseoducode of code MSVC is generating:

Before:
```c
        if ( GenTree::OperIsCopyBlkOp(tree) )
          return (GenTreeOp *)Compiler::fgMorphCopyBlock(this, tree);
        else
          return (GenTreeOp *)Compiler::fgMorphInitBlock(this, tree);
```

The `fgMorphCopyBlock` and `fgMorphInitBlock` calls both inline `MorphInitBlockHelper::Morph` and use some GDVs to avoid some virtual calls and CFG function pointer checks.

After:
```c
        if ( GenTree::OperIsCopyBlkOp(tree) )
        {
          v27.m_initBlock = 0;
          v27.m_blockSize = 0;
          v27.__vftable = (MorphInitBlockHelper_vtbl *)MorphCopyBlockHelper::`vftable';
          v27.m_blockLayout = 0i64;
          memset(&v27.m_dstLclNode, 0, 22);
          v27.m_transformationDecision = Undefined;
          v27.m_result = 0i64;
          v28 = -1;
          v29 = 0i64;
          v30 = 0i64;
          v31 = 0;
          v32 = 0;
          v33 = 0;
          v34 = 0;
        }
        else
        {
          v27.m_initBlock = 1;
          v27.__vftable = (MorphInitBlockHelper_vtbl *)MorphInitBlockHelper::`vftable';
          v27.m_blockSize = 0;
          v27.m_blockLayout = 0i64;
          memset(&v27.m_dstLclNode, 0, 22);
          v27.m_transformationDecision = Undefined;
          v27.m_result = 0i64;
        }
        return (GenTreeOp *)MorphInitBlockHelper::Morph(&v27);
```

The constructors were inlined but we not longer generate specialized versions for `MorphInitBlockHelper::Morph` which uses some virtual calls that go through CFG checks.

AndyAyersMS reviewed (approved):
LGTM

SingleAccretion reviewed (commented):
We also have workarounds in place that can be deleted:

https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/rationalize.cpp#L72-L73 - pretty safe to just delete.

https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/fginline.cpp#L396-L402 - this one has somewhat broad CQ impact so may not be as easy to delete.

Additionally, `gtNewTempAssign` calls `impAssignStruct` which tries to sink commas, which is then worked around in [CSE](https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/optcse.cpp#L3171-L3182) and [return merging](https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/morph.cpp#L13928-L13951). Notionally, it would be possible to get rid of the sinking.

Finally, this fixes #1699.

jakobbotsch commented on src/coreclr/jit/morphblock.cpp:
```diff
@@ -665,6 +547,93 @@ void MorphInitBlockHelper::TryPrimitiveInit()
+//   the created commas. Therefore this function just returns a linked list of
+//   the side effects to be used for that purpose.
+//
+GenTree* MorphInitBlockHelper::EliminateCommas()
```
Yep, I need to handle that.

jakobbotsch commented on src/coreclr/jit/morphblock.cpp:
```diff
@@ -665,6 +547,93 @@ void MorphInitBlockHelper::TryPrimitiveInit()
+    assert(lhs->OperIsUnary() || lhs->OperIsLeaf());
+
+    GenTree* rhs = m_asg->gtGetOp2();
+    if (lhs->OperIsUnary() && ((lhs->gtGetOp1()->gtFlags & GTF_ALL_EFFECT) != 0) && rhs->OperIs(GT_COMMA))
```
Good catch.

jakobbotsch commented:
Updated this to address feedback. It's currently based on #83814 that will need to go in first.

> - pretty safe to just delete.

Thanks! Deleted this one.

> - this one has somewhat broad CQ impact so may not be as easy to delete.

Updated this too, overall an improvement and diffs aren't that big.

> Additionally, gtNewTempAssign calls impAssignStruct which tries to sink commas, which is then worked around in [CSE](https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/optcse.cpp#L3171-L3182) and [return merging](https://github.com/dotnet/runtime/blob/6bcd4aa9cf626d6ee490a7bd35908cefd1603103/src/coreclr/jit/morph.cpp#L13928-L13951). Notionally, it would be possible to get rid of the sinking.

I will leave this as-is for now, the function has some recursive handling right now and it probably would need to be restructured a bit.

jakobbotsch commented on src/coreclr/jit/morphblock.cpp:
```diff
@@ -147,12 +147,18 @@ GenTree* MorphInitBlockHelper::Morph()
     {
-        printf("%s (after):\n", GetHelperName());
-        m_comp->gtDispTree(m_result);
+        m_result = m_comp->gtNewOperNode(GT_COMMA, m_result->TypeGet(), sideEffects, m_result);
```
I ended up reusing the commas by returning a pool of the freed ones from `EliminateCommas`, also linked via their `gtNext` fields. I couldn't find a more elegant way of doing this.

jakobbotsch committed:
JIT: Generalize handling of commas in block morphing

Eliminate commas early in block morphing, before the rest of the
transformation needs to look at it. Do this by extracting their side
effects and adding them on top of the returned result instead. This
allows us to handle arbitrary COMMAs on destination operand in a general
manner.

Prerequisite to #83388.

Fix #1699.

jakobbotsch committed:
Run jit-format

jakobbotsch commented:
@SingleAccretion Was `RewriteSIMDIndir` meant to be for CQ or correctness? We can still see `IND(LCL_VAR_ADDR)` nodes due to forward sub, and I am hitting an assert on this in the backend -- apparently we mark the `LCL_VAR_ADDR` as contained and `emitHandleMemOp` is not prepared to handle that.

jakobbotsch committed:
Revert RewriteSIMDIndir removal

jakobbotsch commented:
I reverted the deletion of `RewriteSIMDIndir`, the backend needs to be taught how to handle `IND<simd12>(LCL_VAR_ADDR/LCL_FLD_ADDR)` with a contained address before it can be deleted. I will open an issue after this is merged.

SingleAccretion commented:
It was meant to be for CQ only (at this point anyway, in the past it was also needed for correctness). Unfortunate that a bug has snuck in the meantime.

jakobbotsch committed:
Reset flags on commas taken from pool

jakobbotsch commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=218439&view=ms.vss-build-web.run-extensions-tab).

Can you please take another look @AndyAyersMS @SingleAccretion?
